solver/llbsolver/solver.go

```diff
@@ -630,42 +617,97 @@ func runCacheExporters(ctx context.Context, exporters []RemoteCacheExporter, j *
 }
 
-func runInlineCacheExporter(ctx context.Context, e exporter.ExporterInstance, inlineExporter *RemoteCacheExporter, j *solver.Job, cached *result.Result[solver.CachedResult]) (map[string][]byte, error) {
-	meta := map[string][]byte{}
+func runInlineCacheExporter(ctx context.Context, e exporter.ExporterInstance, inlineExporter inlineCacheExporter, j *solver.Job, cached *result.Result[solver.CachedResult]) (result *exptypes.InlineCache, err error) {
+	result = &exptypes.InlineCache{
+		Platforms: make(map[string][]byte),
+	}
 	if inlineExporter == nil {
-		return nil, nil
+		return result, nil
 	}
 	if err := inBuilderContext(ctx, j, "preparing layers for inline cache", j.SessionID+"-cache-inline", func(ctx context.Context, _ session.Group) error {
 		if res := cached.Ref; res != nil {
-			dtic, err := inlineCache(ctx, inlineExporter.Exporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
+			dtic, err := inlineCache(ctx, inlineExporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
 			if err != nil {
 				return err
 			}
 			if dtic != nil {
-				meta[exptypes.ExporterInlineCache] = dtic
+				result.Cache = dtic
 			}
 		}
 		for k, res := range cached.Refs {
-			dtic, err := inlineCache(ctx, inlineExporter.Exporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
+			dtic, err := inlineCache(ctx, inlineExporter, res, e.Config().Compression(), session.NewGroup(j.SessionID))
 			if err != nil {
 				return err
 			}
 			if dtic != nil {
-				meta[fmt.Sprintf("%s/%s", exptypes.ExporterInlineCache, k)] = dtic
+				result.Platforms[k] = dtic
 			}
 		}
 		return nil
 	}); err != nil {
 		return nil, err
 	}
-	return meta, nil
+	return result, nil
 }
 
-func splitCacheExporters(exporters []RemoteCacheExporter) (rest []RemoteCacheExporter, inline *RemoteCacheExporter) {
-	rest = make([]RemoteCacheExporter, 0, len(exporters))
+func (s *Solver) runExporters(ctx context.Context, exporters []exporter.ExporterInstance, inlineCacheExporter inlineCacheExporter, job *solver.Job, cached *result.Result[solver.CachedResult], inp *result.Result[cache.ImmutableRef]) (exporterResponse map[string]string, descref exporter.DescriptorReference, err error) {
+	eg, ctx := errgroup.WithContext(ctx)
+	sessionID := job.SessionID
+	resps := make([]map[string]string, len(exporters))
 	for i, exp := range exporters {
-		if _, ok := asInlineCache(exp.Exporter); ok {
-			inline = &exporters[i]
+		func(exp exporter.ExporterInstance, i int) {
+			eg.Go(func() error {
+				if imageExporter, ok := asImageExporter(exp); ok {
+					inlineCache, err := runInlineCacheExporter(ctx, exp, inlineCacheExporter, job, cached)
+					if err != nil {
+						return err
+					}
+					return inBuilderContext(ctx, job, exp.Name(), job.SessionID+"-export", func(ctx context.Context, _ session.Group) (err error) {
+						var dref exporter.DescriptorReference
+						resps[i], dref, err = imageExporter.ExportImage(ctx, inp, *inlineCache, sessionID)
```

Comment:
We can avoid the interface juggling by passing a function that runs the `runInlineCacheExporter` function. Exporters that don't support inline cache just don't need to call that function.

Reply:
The interface helps to determine whether it's necessary to run the inline cache exporter in the first place - if that were a simple function / or placed into `Source` as additional metadata before the corresponding exporter's Export is entered, such optimization wouldn't be possible. I'm ok dropping it if it really simplifies things by a lot in your POV.

Reply:
On the Result type and the need to handle concurrent updates - I see the reasons of simplicity of having a single thread-safe type but I also don't see how to do this with as little changes as possible - this change will be massive, penetrating a lot of places due to potential change of APIs - e.g. instead of allowing direct attribute access, it would have getters/setters instead to exercise greater control (that concerns the metadata of all attributes).

Which also makes me wonder if there won't be a strong case for the 2 type hierarchy - a simple Result not safe for concurrent use (and suiting the majority of use-cases) but also a thread-safe wrapper to compensate for cases where there is a legitimate need to populate the result from multiple goroutines. So far, I could only find a single [place](https://github.com/moby/buildkit/blob/v0.12.0/frontend/dockerui/build.go#L20) where that was necessary. Please let me know if I'm in the wrong here.

With that said, can we also reach some sort of intermediate ground and delay this refactoring to a dedicated issue?